[BUG] cudf.concat() throws concatenating Categorical columns when ignore_index=True

**Describe the bug**
After https://github.com/rapidsai/cudf/pull/5649, `Frame._concat()` throws an error when one of the input frames has a categorical column, but the others do not.

**Steps/Code to reproduce bug**
```python
>>> import cudf
>>> df1 = cudf.DataFrame({'cat': cudf.Series(["one", "two"], dtype="category")})
>>> df2 = cudf.DataFrame({"c": [10, 11, 22, 33, 44, 100]}, index=[7, 8, 9, 10, 11, 20])
>>> cudf.concat([df1, df2], ignore_index=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ptaylor/dev/rapids/cudf/python/cudf/cudf/core/reshape.py", line 193, in concat
    objs, axis=axis, ignore_index=ignore_index, sort=sort
  File "/home/ptaylor/dev/rapids/compose/etc/conda/cuda_10.2/envs/rapids/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/home/ptaylor/dev/rapids/cudf/python/cudf/cudf/core/frame.py", line 281, in _concat
    cast_cols_to_common_dtypes(indices, columns, dtypes, categories)
  File "/home/ptaylor/dev/rapids/cudf/python/cudf/cudf/core/frame.py", line 191, in cast_cols_to_common_dtypes
    cols[idx]
AttributeError: 'StringColumn' object has no attribute 'cat'
```

**Expected behavior**
The dataframes are concatenated without issue.

**Additional context**
Removing these two lines fixes the issue and doesn't lead to other test failures: https://github.com/rapidsai/cudf/blob/ce826c57c8cafb7d78f33873ca8c0c030fa42087/python/cudf/cudf/core/frame.py#L258-L259

However when I try to add a DF with a CategoricalColumn to `test_concat.py::test_concat_empty_dataframes()`, pandas fills the expected categorical column with the value `-1`, but our CategoricalColumn writes `-1` into the _codes_, causing its values to report NaN, and failing the `assert_eq` on line 516 here: https://github.com/rapidsai/cudf/blob/ce826c57c8cafb7d78f33873ca8c0c030fa42087/python/cudf/cudf/tests/test_concat.py#L507-L520
```
E   DataFrame.iloc[:, 0] (column name="cat") values are different (80.0 %)
E   [left]:  [one, two, -1, -1, -1, -1, -1, -1, -1, -1]
E   [right]: [one, two, nan, nan, nan, nan, nan, nan, nan, nan]
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -186,6 +186,7 @@
 - PR #5692 Fix compilation issue with gcc 7.4.0 and CUDA 10.1
 - PR #5693 Add fix missing from PR 5656 to update local docker image to py3.7
 - PR #5703 Small fix for dataframe constructor with cuda array interface objects that don't have `descr` field
+- PR #5719 Fix Frame._concat() with categorical columns
 
 # cuDF 0.14.0 (03 Jun 2020)
 

--- a/python/cudf/cudf/core/frame.py
+++ b/python/cudf/cudf/core/frame.py
@@ -70,9 +70,13 @@ class Frame(libcudf.table.Table):
         # shallow-copy the input DFs in case the same DF instance
         # is concatenated with itself
 
-        some_empty = False
+        # flag to indicate at least one empty input frame also has an index
+        empty_has_index = False
+        # length of output frame's RangeIndex if all input frames are empty,
+        # and at least one has an index
         result_index_length = 0
-        empty_counter = 0
+        # the number of empty input frames
+        num_empty_input_frames = 0
 
         for i, obj in enumerate(objs):
             objs[i] = obj.copy(deep=False)
@@ -86,20 +90,9 @@ class Frame(libcudf.table.Table):
                 # variable. Detailed explanation of second case before
                 # allocation of `columns` variable below.
                 if obj.empty:
-                    empty_counter += 1
-                    if len(obj) > 0:
-                        some_empty = True
-                        result_index_length += len(obj)
-
-        # some_empty indicates there are dataframes in obj
-        # which are empty and have index.
-
-        # all_empty indicates all dataframes in obj which
-        # are empty and at least one of them have index.
-        if empty_counter == len(objs) and some_empty:
-            all_empty = True
-        else:
-            all_empty = False
+                    num_empty_input_frames += 1
+                    result_index_length += len(obj)
+                    empty_has_index = empty_has_index or len(obj) > 0
 
         from cudf.core.column.column import (
             build_categorical_column,
@@ -113,7 +106,7 @@ class Frame(libcudf.table.Table):
             dtypes = dict()
             # A mapping of {idx: [...columns]}, where `[...columns]`
             # is a list of columns with at least one valid value for each
-            # column name across all input dataframes
+            # column name across all input frames
             non_null_columns = dict()
             for idx in col_idxs:
                 for cols in list_of_columns:
@@ -136,7 +129,7 @@ class Frame(libcudf.table.Table):
         def find_common_dtypes_and_categories(non_null_columns, dtypes):
             # A mapping of {idx: categories}, where `categories` is a
             # column of all the unique categorical values from each
-            # categorical column across all input dataframes
+            # categorical column across all input frames
             categories = dict()
             for idx, cols in non_null_columns.items():
                 # default to the first non-null dtype
@@ -226,7 +219,7 @@ class Frame(libcudf.table.Table):
 
         # Combine the index and table columns for each Frame into a list of
         # [...index_cols, ...table_cols].
-        # 
+        #
         # If any of the input frames have a non-empty index, include these
         # columns in the list of columns to concatenate, even if the input
         # frames are empty and `ignore_index=True`.
@@ -234,7 +227,7 @@ class Frame(libcudf.table.Table):
         columns = [
             (
                 []
-                if (ignore_index and not some_empty)
+                if (ignore_index and not empty_has_index)
                 else list(f._index._data.columns)
             )
             + [f._data[name] if name in f._data else None for name in names]
@@ -292,10 +285,10 @@ class Frame(libcudf.table.Table):
             libcudf.concat.concat_tables(tables, ignore_index=ignore_index)
         )
 
-        if all_empty:
-            # When all the dataframes in objs are empty and either
-            # one of them have index and ignore_index is True, create
-            # a new RangeIndex and assign it to the result.
+        # If ignore_index is True, all input frames are empty, and at
+        # least one input frame has an index, assign a new RangeIndex
+        # to the result frame.
+        if empty_has_index and num_empty_input_frames == len(objs):
             out._index = cudf.RangeIndex(result_index_length)
 
         # Reassign the categories for any categorical table cols

--- a/python/cudf/cudf/core/frame.py
+++ b/python/cudf/cudf/core/frame.py
@@ -252,14 +252,7 @@ class Frame(libcudf.table.Table):
                 if (ignore_index and not some_empty)
                 else list(f._index._data.columns)
             )
-            + [
-                f._data[name]
-                if name in f._data
-                else column_empty(len(f), masked=True)
-                if (ignore_index and not some_empty)
-                else None
-                for name in names
-            ]
+            + [f._data[name] if name in f._data else None for name in names]
             for i, f in enumerate(objs)
         ]
 

--- a/python/cudf/cudf/core/frame.py
+++ b/python/cudf/cudf/core/frame.py
@@ -224,27 +224,12 @@ class Frame(libcudf.table.Table):
         except TypeError:
             names = list(names)
 
-        # Combine the index and table columns for each Frame into a
-        # list of [...index_cols, ...table_cols]. If a table is
-        # missing a column:
-        # 1. when ignore_index is True and some_empty is False(which
-        # means there is no dataframe which is empty and has an index in
-        # objs) an empty column of length of the respective dataframe
-        # is added to the slot.
-        # 2. when ignore_index is False or some_empty is True list
-        # will have None in the slot instead
-
-        # When ignore_index is True, and if at least one of the
-        # dataframe in objs is empty and has and index set:
-        #
-        # 1. Do not ignore the index for index_cols being
-        # populated in columns[0]. Instead we will be ignoring the
-        # index in libcudf.concat.concat_tables api call. This is
-        # necessary for cast_cols_to_common_dtypes api.
-        # 2. Create an empty column and populate in columns[1] as
-        # index will be ignored by libcudf.concat.concat_tables
-        # and the corresponding empty column will be concatenated
-        # correctly.
+        # Combine the index and table columns for each Frame into a list of
+        # [...index_cols, ...table_cols].
+        # 
+        # If any of the input frames have a non-empty index, include these
+        # columns in the list of columns to concatenate, even if the input
+        # frames are empty and `ignore_index=True`.
 
         columns = [
             (

--- a/python/cudf/cudf/tests/test_concat.py
+++ b/python/cudf/cudf/tests/test_concat.py
@@ -8,6 +8,7 @@ import pytest
 
 import cudf as gd
 from cudf.tests.utils import assert_eq
+from cudf.utils.dtypes import is_categorical_dtype
 
 
 def make_frames(index=None, nulls="none"):
@@ -481,6 +482,7 @@ def test_concat_series_dataframe_input_str(objs):
         pd.DataFrame({"l": [10]}),
         pd.DataFrame({"l": [10]}, index=[200]),
         pd.DataFrame([], index=[100]),
+        pd.DataFrame({"cat": pd.Series(["one", "two"], dtype="category")}),
     ],
 )
 @pytest.mark.parametrize(
@@ -513,6 +515,10 @@ def test_concat_empty_dataframes(df, other, ignore_index):
     expected = pd.concat(other_pd, ignore_index=ignore_index)
     actual = gd.concat(other_gd, ignore_index=ignore_index)
     if expected.shape != df.shape:
+        for key, col in actual[actual.columns].iteritems():
+            if is_categorical_dtype(col.dtype):
+                expected[key] = expected[key].fillna("-1")
+                actual[key] = col.astype("str").fillna("-1")
         assert_eq(expected.fillna(-1), actual.fillna(-1), check_dtype=False)
     else:
         assert_eq(

--- a/python/cudf/cudf/tests/test_concat.py
+++ b/python/cudf/cudf/tests/test_concat.py
@@ -501,8 +501,13 @@ def test_concat_series_dataframe_input_str(objs):
             pd.DataFrame({"f": [10.2, 11.2332, 0.22, 3.3, 44.23, 10.0]}),
             pd.DataFrame({"l": [10]}),
             pd.DataFrame({"l": [10]}, index=[200]),
+            pd.DataFrame({"cat": pd.Series(["two", "three"], dtype="category")}),
         ],
-        [pd.DataFrame([]), pd.DataFrame([], index=[100])],
+        [
+            pd.DataFrame([]),
+            pd.DataFrame([], index=[100]),
+            pd.DataFrame({"cat": pd.Series(["two", "three"], dtype="category")}),
+        ],
     ],
 )
 @pytest.mark.parametrize("ignore_index", [True, False])

--- a/python/cudf/cudf/tests/test_concat.py
+++ b/python/cudf/cudf/tests/test_concat.py
@@ -501,12 +501,16 @@ def test_concat_series_dataframe_input_str(objs):
             pd.DataFrame({"f": [10.2, 11.2332, 0.22, 3.3, 44.23, 10.0]}),
             pd.DataFrame({"l": [10]}),
             pd.DataFrame({"l": [10]}, index=[200]),
-            pd.DataFrame({"cat": pd.Series(["two", "three"], dtype="category")}),
+            pd.DataFrame(
+                {"cat": pd.Series(["two", "three"], dtype="category")}
+            ),
         ],
         [
             pd.DataFrame([]),
             pd.DataFrame([], index=[100]),
-            pd.DataFrame({"cat": pd.Series(["two", "three"], dtype="category")}),
+            pd.DataFrame(
+                {"cat": pd.Series(["two", "three"], dtype="category")}
+            ),
         ],
     ],
 )

